File: apps/forja/lib/shared/extractors/core/stream_crypto.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter/foundation.dart' show visibleForTesting;

/// STREAMCRYPTO (`enc=2`) — seed + media id → XOR keystream → `mvm1` JSON.
///
/// Same algorithm as the player JS and AniWorld's Python port. Used by every
/// Forja path that hits this player family (Videasy provider, VidSrc.sbs 4K
/// `player.videasy.*` nested embeds, …) — not a Videasy-only cipher.
abstract final class StreamCrypto {
  static const _mask = 0xFFFFFFFF;
  static const _golden = 2654435769; // 0x9E3779B9
  static const _magic = [109, 118, 109, 49]; // mvm1

  /// True when [url] is the enc=2 player (not Cinesrc / nxsha / …).
  static bool isPlayerUrl(String url, {String? configuredOrigin}) {
    final host = Uri.tryParse(url.trim())?.host.toLowerCase() ?? '';
    if (host.isEmpty) return false;
    if (_isVideasyFamilyHost(host)) return true;
    final extra = Uri.tryParse(configuredOrigin?.trim() ?? '')?.host
            .toLowerCase() ??
        '';
    return extra.isNotEmpty && host == extra;
  }

  static bool _isVideasyFamilyHost(String host) {
    return host == 'videasy.to' ||
        host == 'videasy.net' ||
        host.endsWith('.videasy.to') ||
        host.endsWith('.videasy.net');
  }

  /// Decrypt an `enc=2` body into UTF-8 JSON (no `mvm1` prefix).
  ///
  /// Throws [FormatException] on bad magic / garbage.
  static String decrypt(String payload, String seed, String mediaId) {
    final data = _b64urlDecode(payload);
    if (data.length < _magic.length) {
      throw const FormatException('STREAMCRYPTO: payload too short');
    }
    final id = int.tryParse(mediaId.trim());
    if (id == null) {
      throw FormatException('STREAMCRYPTO: invalid media id $mediaId');
    }
    final ks = _keystream(seed, id, data.length);
    for (var i = 0; i < data.length; i++) {
      data[i] ^= ks[i];
    }
    for (var j = 0; j < _magic.length; j++) {
      if (data[j] != _magic[j]) {
        throw const FormatException(
          'STREAMCRYPTO: bad seed or tampered payload',
        );
      }
    }
    return utf8.decode(data.sublist(_magic.length));
  }

  /// Inverse of [decrypt] for unit round-trips. Not used in extract.
  @visibleForTesting
  static String encryptForTest(String json, String seed, String mediaId) {
    final id = int.parse(mediaId);
    final plain = Uint8List.fromList([..._magic, ...utf8.encode(json)]);
    final ks = _keystream(seed, id, plain.length);
    for (var i = 0; i < plain.length; i++) {
      plain[i] ^= ks[i];
    }
    return base64Url.encode(plain).replaceAll('=', '');
  }

  static int _imul(int a, int b) => (a * b) & _mask;

  static int _f(int e) {
    e &= _mask;
    e ^= e >> 16;
    e = _imul(e, 2246822507);
    e ^= e >> 13;
    e = _imul(e, 3266489909);
    e ^= e >> 16;
    return e & _mask;
  }

  static int _rotl(int e, int t) {
    e &= _mask;
    t &= 31;
    if (t == 0) return e;
    return ((e << t) | (e >> (32 - t))) & _mask;
  }

  static int _fnvF(String text) {
    var t = 2166136261;
    for (final code in text.codeUnits) {
      t = _imul(t ^ code, 16777619);
    }
    return _f(t);
  }

  static ({Map<int, int> state, int acc}) _keySchedule(
    String seed,
    int mediaId,
  ) {
    var n = _f(_fnvF(seed) ^ _f((mediaId & _mask) ^ _golden));
    final state = <int, int>{};
    for (var e = 0; e < 8; e++) {
      final idx = n % 61;
      n = _rotl((n + _golden) & _mask, 7 + (e & 7));
      state[idx] = (n ^ _f(n)) & _mask;
      n = _f((n + idx) & _mask);
    }
    final acc = _f(2779096485 ^ n);
    return (state: state, acc: acc);
  }

  static Uint8List _keystream(String seed, int mediaId, int length) {
    final sched = _keySchedule(seed, mediaId);
    final state = Map<int, int>.from(sched.state);
    var acc = sched.acc;
    final out = Uint8List(length);
    var pos = 0;
    var counter = 0;
    while (pos < length) {
      final a = acc;
      final i = a % 61;
      final mask = state.containsKey(i) ? _mask : 0;
      final low = (state[i] ?? 0) & _mask;
      final mixed = (low ^ _imul(_golden, counter + 1)) & _mask;
      var c = ((a ^ mixed) | (a & mixed & mask)) & _mask;
      c = (_rotl((c + a) & _mask, i & 31) ^ _rotl(a, _imul(i, 7) & 31)) &
          _mask;
      acc = _f((c + _golden) & _mask);
      state[i] = acc & _mask;
      counter++;
      final val = acc;
      out[pos++] = val & 255;
      if (pos < length) out[pos++] = (val >> 8) & 255;
      if (pos < length) out[pos++] = (val >> 16) & 255;
      if (pos < length) out[pos++] = (val >> 24) & 255;
    }
    return out;
  }

  static Uint8List _b64urlDecode(String text) {
    var t = text.trim().replaceAll('-', '+').replaceAll('_', '/');
    if (t.isEmpty) {
      throw const FormatException('STREAMCRYPTO: empty payload');
    }
    final pad = (4 - t.length % 4) % 4;
    t += '=' * pad;
    try {
      return Uint8List.fromList(base64.decode(t));
    } on FormatException {
      throw const FormatException('STREAMCRYPTO: invalid payload');
    }
  }
}

File: crates/engine-js/src/stream_crypto.rs
//! STREAMCRYPTO (`enc=2`) — port of Dart `StreamCrypto.decrypt`.

use base64::{engine::general_purpose::STANDARD, Engine as _};

const MASK: u32 = 0xFFFF_FFFF;
const GOLDEN: u32 = 0x9E37_79B9;
const MAGIC: [u8; 4] = [109, 118, 109, 49]; // mvm1

#[derive(Debug, thiserror::Error)]
pub enum StreamCryptoError {
    #[error("STREAMCRYPTO: payload too short")]
    TooShort,
    #[error("STREAMCRYPTO: invalid media id")]
    BadMediaId,
    #[error("STREAMCRYPTO: empty payload")]
    Empty,
    #[error("STREAMCRYPTO: invalid payload")]
    InvalidB64,
    #[error("STREAMCRYPTO: bad seed or tampered payload")]
    BadMagic,
    #[error("STREAMCRYPTO: utf8")]
    Utf8,
}

fn imul(a: u32, b: u32) -> u32 {
    a.wrapping_mul(b)
}

fn f(mut e: u32) -> u32 {
    e ^= e >> 16;
    e = imul(e, 2246822507);
    e ^= e >> 13;
    e = imul(e, 3266489909);
    e ^= e >> 16;
    e
}

fn rotl(e: u32, t: u32) -> u32 {
    let t = t & 31;
    if t == 0 {
        e
    } else {
        e.rotate_left(t)
    }
}

fn fnv_f(text: &str) -> u32 {
    let mut t: u32 = 2166136261;
    for code in text.chars() {
        t = imul(t ^ (code as u32), 16777619);
    }
    f(t)
}

fn key_schedule(seed: &str, media_id: u32) -> (std::collections::HashMap<u32, u32>, u32) {
    let mut n = f(fnv_f(seed) ^ f((media_id & MASK) ^ GOLDEN));
    let mut state = std::collections::HashMap::new();
    for e in 0..8u32 {
        let idx = n % 61;
        n = rotl(n.wrapping_add(GOLDEN), 7 + (e & 7));
        state.insert(idx, n ^ f(n));
        n = f(n.wrapping_add(idx));
    }
    let acc = f(2779096485 ^ n);
    (state, acc)
}

fn keystream(seed: &str, media_id: u32, length: usize) -> Vec<u8> {
    let (mut state, mut acc) = key_schedule(seed, media_id);
    let mut out = vec![0u8; length];
    let mut pos = 0usize;
    let mut counter = 0u32;
    while pos < length {
        let a = acc;
        let i = a % 61;
        let mask = if state.contains_key(&i) { MASK } else { 0 };
        let low = *state.get(&i).unwrap_or(&0);
        let mixed = low ^ imul(GOLDEN, counter + 1);
        let mut c = (a ^ mixed) | (a & mixed & mask);
        c = rotl(c.wrapping_add(a), i & 31) ^ rotl(a, imul(i, 7) & 31);
        acc = f(c.wrapping_add(GOLDEN));
        state.insert(i, acc);
        counter = counter.wrapping_add(1);
        let val = acc;
        out[pos] = (val & 255) as u8;
        pos += 1;
        if pos < length {
            out[pos] = ((val >> 8) & 255) as u8;
            pos += 1;
        }
        if pos < length {
            out[pos] = ((val >> 16) & 255) as u8;
            pos += 1;
        }
        if pos < length {
            out[pos] = ((val >> 24) & 255) as u8;
            pos += 1;
        }
    }
    out
}

fn b64url_decode(text: &str) -> Result<Vec<u8>, StreamCryptoError> {
    let mut t = text.trim().replace('-', "+").replace('_', "/");
    if t.is_empty() {
        return Err(StreamCryptoError::Empty);
    }
    let pad = (4 - t.len() % 4) % 4;
    t.push_str(&"=".repeat(pad));
    STANDARD
        .decode(t.as_bytes())
        .map_err(|_| StreamCryptoError::InvalidB64)
}

/// Decrypt an `enc=2` body into UTF-8 JSON (no `mvm1` prefix).
pub fn decrypt(payload: &str, seed: &str, media_id: &str) -> Result<String, StreamCryptoError> {
    let mut data = b64url_decode(payload)?;
    if data.len() < MAGIC.len() {
        return Err(StreamCryptoError::TooShort);
    }
    let id: u32 = media_id
        .trim()
        .parse()
        .map_err(|_| StreamCryptoError::BadMediaId)?;
    let ks = keystream(seed, id, data.len());
    for i in 0..data.len() {
        data[i] ^= ks[i];
    }
    if data[..MAGIC.len()] != MAGIC {
        return Err(StreamCryptoError::BadMagic);
    }
    String::from_utf8(data[MAGIC.len()..].to_vec()).map_err(|_| StreamCryptoError::Utf8)
}

/// Inverse of [decrypt] for unit round-trips.
#[cfg(test)]
pub fn encrypt_for_test(json: &str, seed: &str, media_id: &str) -> Result<String, StreamCryptoError> {
    let id: u32 = media_id.parse().map_err(|_| StreamCryptoError::BadMediaId)?;
    let mut plain = Vec::with_capacity(MAGIC.len() + json.len());
    plain.extend_from_slice(&MAGIC);
    plain.extend_from_slice(json.as_bytes());
    let ks = keystream(seed, id, plain.len());
    for i in 0..plain.len() {
        plain[i] ^= ks[i];
    }
    Ok(base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(plain))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let seed = "test-seed-abc";
        let media = "94997";
        let json = r#"{"sources":[{"url":"https://x/a.m3u8","quality":"1080p"}]}"#;
        let payload = encrypt_for_test(json, seed, media).unwrap();
        let out = decrypt(&payload, seed, media).unwrap();
        assert_eq!(out, json);
    }
}
